### PR TITLE
Honor end_of_line editorconfig option when splitting strings

### LINKED ISSUE #41322: C#, Editor, 'Split string literals on enter' not honoring EOL convention.
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/content/problem/889591/c-editor-split-string-literals-on-enter-not-honori.html)._

---
Repro:
Create or modify .editorconfig for a Solution with
```
[*.cs]
end_of_line = lf
```

Make sure the option 'Split string literals on enter' is enabled.

From a project in the Solution, Open a C# source file containing a string.

At the very bottom-right of the editor view for that file, the area displaying EOL sequence, make sure it reads "LF". If not, click it and select LF.

Place cursor in the middle of a string in the C# source file.

Hit return.

Actual:
Note that EOL sequence now, in error, reads "MIXED". The operation has unconditionally introduced CRLF.

Expected:
Honor the .editorconfig setting, and the current source file's line endings (which if consistent, i.e. not MIXED, should take precedence for this operation).

---
#### Visual Studio Feedback System on 1/19/2020, 00:31 AM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

---
### Original Solutions
(no solutions)

### PR DESCRIPTION
**Pull request by 06needhamt**
Fixes #41322 

This PR ensures that the string literal splitter honours the value in the `end_of_line` editorconfig option.

I have also added a test for this but there is an issue where all the line endings on test markup revers to CRLF even when I insert a string with LF endings. I am not sure how to resolve this problem.

### CODE REVIEW

File: src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.cs [hunk 1]
```diff
@@ -136,16 +136,7 @@ private bool SplitString(ITextView textView, ITextBuffer subjectBuffer, Snapshot
...
-            for (int i = line.Start; i < caretPosition; i++)
-            {
-                if (snapshot[i] == '"')
-                {
-                    return true;
-                }
-            }
-
-            return false;
+            return line.Snapshot.GetText().Contains("\"");
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.cs]: This allocates. Please undo

File: src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs [hunk 2]
```diff
@@ -46,6 +44,27 @@ private abstract class StringSplitter
...
                 _indentStyle = indentStyle;
                 CancellationToken = cancellationToken;
+
+                var options = document.GetOptionsAsync().Result;
+                var newLineKey = new OptionKey(NewLine, LanguageNames.CSharp);
+                var newLineValue = (string)options.GetOption(newLineKey);
+
+                // Check if there is a line ending specified in the editorconfig
+                // and ensure we honour it. 
+                if (!string.IsNullOrEmpty(newLineValue) && newLineValue == "\n")
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs]: IsNullOrEmpty check is not needed

File: src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs [hunk 3]
```diff
@@ -46,6 +44,27 @@ private abstract class StringSplitter
...
+                   SyntaxKind.PlusToken,
+                   SyntaxFactory.TriviaList(SyntaxFactory.ElasticLineFeed));
+                }
+                else
+                {
+                    PlusNewLineToken = SyntaxFactory.Token(
+                    leading: default,
+                    SyntaxKind.PlusToken,
+                    SyntaxFactory.TriviaList(SyntaxFactory.ElasticCarriageReturnLineFeed));
+                }
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs]: Can simplify all of this to just make the token but only conditionally switch the type of elastic trivia to add

File: src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.cs (hunk 1 above)
Comment (06needhamt) [reply src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.cs]: Done

File: src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs (hunk 2 above)
Comment (06needhamt) [reply src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs]: Removed

File: src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs (hunk 3 above)
Comment (06needhamt) [reply src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs]: Done

File: src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs [hunk 4]
```diff
@@ -46,6 +44,17 @@ private abstract class StringSplitter
...
+                var options = document.GetOptionsAsync().Result;
+                var newLineKey = new OptionKey(NewLine, LanguageNames.CSharp);
+                var newLineValue = (string)options.GetOption(newLineKey);
+
+                // Check if there is a line ending specified in the editorconfig
+                // and ensure we honour it. 
+                PlusNewLineToken = SyntaxFactory.Token(
+                leading: default,
+                SyntaxKind.PlusToken,
+                newLineValue == "\n" ? SyntaxFactory.TriviaList(SyntaxFactory.ElasticLineFeed) : SyntaxFactory.TriviaList(SyntaxFactory.ElasticCarriageReturnLineFeed));
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs]: indent these three lines one leve.

File: src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs [hunk 5]
```diff
@@ -46,6 +44,17 @@ private abstract class StringSplitter
                 TabSize = tabSize;
                 _indentStyle = indentStyle;
                 CancellationToken = cancellationToken;
+
+                var options = document.GetOptionsAsync().Result;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs]: SplitStringLiteralCommandHandler already handles getting and using options.  so can we move this to that code and then have it pass in the appropriate newline trivia into the splitter.

File: src/EditorFeatures/CSharpTest/SplitStringLiteral/SplitStringLiteralCommandHandlerTests.cs [hunk 6]
```diff
@@ -37,18 +37,25 @@ public class SplitStringLiteralCommandHandlerTests
...
+                .WithChangedOption(NewLine, LanguageNames.CSharp, lineEnding)));
 
             if (useTabs && expectedOutputMarkup != null)
             {
                 Assert.Contains("\t", expectedOutputMarkup);
             }
 
+            if (lineEnding == "\n" && expectedOutputMarkup != null)
+            {
+                Assert.DoesNotContain("\r\n", expectedOutputMarkup);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/SplitStringLiteral/SplitStringLiteralCommandHandlerTests.cs]: can you assert this for the inputMarkup as well?  And also the inverse (i.e. if the lineEnding is `\r\n` that we don't have naked `\n`s in the markup (can be checked with regexes).

File: src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs (hunk 4 above)
Comment (06needhamt) [reply src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs]: Done

File: src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs (hunk 5 above)
Comment (06needhamt) [reply src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs]: Done

File: src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs [hunk 7]
```diff
@@ -45,13 +45,20 @@ private abstract class StringSplitter
                 UseTabs = useTabs;
                 TabSize = tabSize;
                 _indentStyle = indentStyle;
+                NewLineTrivia = newLineTrivia;
                 CancellationToken = cancellationToken;
+
+                PlusNewLineToken = SyntaxFactory.Token(
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs]: nit: could easily be one line.

File: src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs [hunk 8]
```diff
@@ -10,24 +10,23 @@
...
-                SyntaxFactory.TriviaList(SyntaxFactory.ElasticCarriageReturnLineFeed));
+            protected readonly SyntaxToken PlusNewLineToken;
 
             protected readonly Document Document;
             protected readonly int CursorPosition;
             protected readonly SourceText SourceText;
             protected readonly SyntaxNode Root;
             protected readonly int TabSize;
             protected readonly bool UseTabs;
+            protected readonly SyntaxTriviaList NewLineTrivia;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs]: this field seems to be unnecessary (i can't see any reads of it anywhere).

File: src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.cs [hunk 9]
```diff
@@ -154,13 +155,17 @@ private bool LineContainsQuote(ITextSnapshotLine line, int caretPosition)
             var useTabs = options.GetOption(FormattingOptions.UseTabs);
             var tabSize = options.GetOption(FormattingOptions.TabSize);
             var indentStyle = options.GetOption(FormattingOptions.SmartIndent, LanguageNames.CSharp);
+            var lineEnding = options.GetOption(FormattingOptions.NewLine, LanguageNames.CSharp);
 
             var root = document.GetSyntaxRootSynchronously(cancellationToken);
             var sourceText = root.SyntaxTree.GetText(cancellationToken);
+            var newLineTrivia = lineEnding == "\n" ? SyntaxFactory.TriviaList(SyntaxFactory.ElasticLineFeed) : SyntaxFactory.TriviaList(SyntaxFactory.ElasticCarriageReturnLineFeed);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.cs]:
```suggestion
            var newLineTrivia = SyntaxFactory.TriviaList(lineEnding == "\n" ? SyntaxFactory.ElasticLineFeed : SyntaxFactory.ElasticCarriageReturnLineFeed);
```

Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.cs]: @jasonmalinowski is this the right way for features to determien appropriate line ending?  I seem to recall you had a PR where there was some core service that could be used that would figure this out based both on settings and heuristics about the file in question.  DId that go through?  Should that be used here instead?

Thanks!

File: src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs (hunk 7 above)
Comment (06needhamt) [reply src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs]: Done

File: src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs (hunk 8 above)
Comment (06needhamt) [reply src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs]: I’ve removed it.

File: src/EditorFeatures/CSharpTest/SplitStringLiteral/SplitStringLiteralCommandHandlerTests.cs (hunk 6 above)
Comment (06needhamt) [reply src/EditorFeatures/CSharpTest/SplitStringLiteral/SplitStringLiteralCommandHandlerTests.cs]: Done

**Comment by 06needhamt** [comment]
@CyrusNajmabadi I’ve addressed all of your comments, however the tests are still failing as I’m not sure how to force line endings in test code. Can you help with this?

**Comment by CyrusNajmabadi** [comment]
Lookin!

**Comment by CyrusNajmabadi** [comment]
FYI, this is failing in TestHonourEndOfLineOption, which i'm guessing you added?

**Comment by 06needhamt** [comment]
>FYI, this is failing in TestHonourEndOfLineOption, which i'm guessing you added?  I can try to fix.  But is it not something you can debug/validate locally?

@CyrusNajmabadi The issue is that the line endings are CRLF in verbatim strings and I need to force the line endings to be LF for the TestHonourEndOfLineOption test but I’m not sure how to do it.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by CyrusNajmabadi** [comment]
@06needhamt you can merge in https://github.com/CyrusNajmabadi/roslyn/tree/patch-41322 to have this working.  Thanks! :)

**Comment by 06needhamt** [comment]
@CyrusNajmabadi I've merged it in, thanks for the help!

**Comment by CyrusNajmabadi** [comment]
yup yup!  @sharwell This is ready for IDE review.  Thanks! :)

**Comment by 06needhamt** [comment]
Ping @sharwell

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by 06needhamt** [comment]
Ping @roslyn-ide @sharwell for review

**Comment by 06needhamt** [comment]
Ping @sharwell for review

File: src/EditorFeatures/CSharpTest/SplitStringLiteral/SplitStringLiteralCommandHandlerTests.cs [hunk 10]
```diff
@@ -891,5 +915,29 @@ void M()
...
+    }
+}".Replace("\r\n", "\n"),
+@"class C
+{
+    void M()
+    {
+        var v = ""Hello "" +
+            ""[||]World!"";
+    }
+}".Replace("\r\n", "\n"), lineEnding: "\n");
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/SplitStringLiteral/SplitStringLiteralCommandHandlerTests.cs]: can you add cases for `$` `@` and `@$` strings?

File: src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs [hunk 11]
```diff
@@ -10,18 +10,15 @@
...
     internal partial class SplitStringLiteralCommandHandler
     {
         private abstract class StringSplitter
         {
             protected static readonly SyntaxAnnotation RightNodeAnnotation = new SyntaxAnnotation();
 
-            protected static readonly SyntaxToken PlusNewLineToken = SyntaxFactory.Token(
-                leading: default,
-                SyntaxKind.PlusToken,
-                SyntaxFactory.TriviaList(SyntaxFactory.ElasticCarriageReturnLineFeed));
```
Comment (sharwell) [inline src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs:23]: 💭 I would expect an elastic trivia to get updated to have the correct form, without needing to separately track the newline configuration inside the fix. I wonder if we can get the `CodeAction` infrastructure to do this for us.

**Comment by sharwell** [comment]
This has come up a few times in previous features. The intent of elastic trivia is avoiding the requirement that each individual feature need to concern itself with details like this. I verified that this feature is already using elastic trivia that is expected to always apply itself with the correct final form:

https://github.com/dotnet/roslyn/blob/5dab61aa59868a5e017f13fd0c89434e97ac21c1/src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs#L26-L29

If the feature is failing to use the correct line endings, the changes need to be made where the formatter handles elastic trivia, and not inside this feature.

Note that tests covering this scenario would of course be welcome as an independent change.

**Pull request closed without merging**